connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -0,0 +1,535 @@
+                    // unaccounted resources
+                    delay = calculateDelay(now);
+                } else {
+                    // This means scheduledRebalance == 0
+                    candidateWorkersForReassignment.addAll(completeWorkerAssignment.stream()
+                            .filter(WorkerLoad::isEmpty)
+                            .map(WorkerLoad::worker)
+                            .collect(Collectors.toSet()));
+                    // we can also extra the current minimum delay from the group, to make
+                    // independent of consecutive leader failures.
+                    delay = maxDelay;
+                }
+                scheduledRebalance = now + delay;
+            }
+        }
+    }
+
+    /**
+     * TODO: describe algorithm
+     *
+     * @param activeAssignments
+     * @param completeWorkerAssignment
+     * @return
+     */
+    private Map<String, ConnectorsAndTasks> performTaskRevocation(ConnectorsAndTasks activeAssignments,
+                                                                  Collection<WorkerLoad> completeWorkerAssignment) {
+        int totalActiveConnectorsNum = activeAssignments.connectors().size();
+        int totalActiveTasksNum = activeAssignments.tasks().size();
+        Collection<WorkerLoad> existingWorkers = completeWorkerAssignment.stream()
+                .filter(wl -> wl.size() > 0)
+                .collect(Collectors.toList());
+        int existingWorkersNum = existingWorkers.size();
+        int totalWorkersNum = completeWorkerAssignment.size();
+        int newWorkersNum = totalWorkersNum - existingWorkersNum;
+
+        Map<String, ConnectorsAndTasks> revoking = new HashMap<>();
+        if (!(newWorkersNum > 0 && existingWorkersNum > 0)) {
+            log.debug("Business as usual existing {} new {} total {}", existingWorkersNum,
+                    newWorkersNum, totalWorkersNum
+            );
+
+            if (log.isDebugEnabled()) {
+                completeWorkerAssignment.forEach(wl -> log.debug(
+                        "Current plan worker {} connectors {} tasks {}",
+                        wl.worker(),
+                        wl.connectorsSize(),
+                        wl.tasksSize()));
+            }
+            return revoking;
+        }
+
+        log.debug("Free workers {} loaded workers {} total workers {}",
+                newWorkersNum, existingWorkersNum, totalWorkersNum);
+
+        int floorConnectors = totalActiveConnectorsNum / totalWorkersNum;
+        int floorTasks = totalActiveTasksNum / totalWorkersNum;
+
+        log.debug("Old connectors per worker {}", totalActiveConnectorsNum / existingWorkersNum);
+        log.debug("New connectors per worker {}", totalActiveConnectorsNum / totalWorkersNum);
+        log.debug("Old tasks per worker {}", totalActiveTasksNum / existingWorkersNum);
+        log.debug("New tasks per worker {}", totalActiveTasksNum / totalWorkersNum);
```

Comment:
Isn't it possible that some of these are 0 due to integer division truncating? Seems like all 4 of these lines could be replaced with something like:
```
log.debug("Assigning {} connectors and {} tasks to {} workers, was previously on {} workers",
        totalActiveConnectorsNum, totalActiveTasksNum, totalWorkersNum, existingWorkersNum);
```

Reply:
This block now reads:
```java

        // We have at least one worker assignment (the leader itself) so totalWorkersNum can't be 0
        log.debug("Previous rounded down (floor) average number of connectors per worker {}", totalActiveConnectorsNum / existingWorkersNum);
        int floorConnectors = totalActiveConnectorsNum / totalWorkersNum;
        log.debug("New rounded down (floor) average number of connectors per worker {}", floorConnectors);

        log.debug("Previous rounded down (floor) average number of tasks per worker {}", totalActiveTasksNum / existingWorkersNum);
        int floorTasks = totalActiveTasksNum / totalWorkersNum;
        log.debug("New rounded down (floor) average number of tasks per worker {}", floorTasks);
```

We shouldn't mind these averages being 0. This means there's not much to revoke.